src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs

```diff
@@ -142,14 +143,33 @@ public void AddNewSearchParameters(IReadOnlyCollection<ITypedElement> searchPara
 
         private void CalculateSearchParameterHash()
         {
-            foreach (KeyValuePair<string, IDictionary<string, SearchParameterInfo>> kvp in TypeLookup)
+            foreach (string resourceName in TypeLookup.Keys)
             {
-                string searchParamHash = SearchHelperUtilities.CalculateSearchParameterHash(kvp.Value.Values);
+                string searchParamHash = SearchHelperUtilities.CalculateSearchParameterHash(TypeLookup[resourceName].Values);
                 _resourceTypeSearchParameterHashMap.AddOrUpdate(
-                    kvp.Key,
+                    resourceName,
                     searchParamHash,
                     (resourceType, existingValue) => searchParamHash);
             }
         }
+
+        public void DeleteSearchParameter(ITypedElement searchParam)
+        {
+            var searchParamWrapper = new SearchParameterWrapper(searchParam);
+            SearchParameterInfo searchParameterInfo = null;
+
+            if (!UrlLookup.TryRemove(new Uri(searchParamWrapper.Url), out searchParameterInfo))
+            {
+                throw new Exception(string.Format(Resources.CustomSearchParameterNotfound, searchParamWrapper.Url));
+            }
+
+            var allResourceTypes = searchParameterInfo.TargetResourceTypes.Union(searchParameterInfo.BaseResourceTypes);
+            foreach (var resourceType in allResourceTypes)
+            {
+                TypeLookup[resourceType].TryRemove(searchParameterInfo.Name, out var removedParam);
```

Comment:
>TryRemove [](start = 41, length = 9)

if all `TryRemove` return false, it's probably doesn't make sense to do `CalculateSearchParameterHash`

Reply:
This is true, but if we passed the first attempt to remove a SearchParameter from the UrlLookup, we should have at least one entry in the TypeLookup to remove.